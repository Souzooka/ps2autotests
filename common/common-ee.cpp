#include <stdio.h>
#include <stdarg.h>
#include <string.h>
#include <kernel.h>

#define TEST_THREAD_PRIORITY   0x40
#define TEST_THREAD_STACK_SIZE 0x8000

char schedfBuffer[65536];
unsigned int schedfBufferPos = 0;

void schedf(const char *format, ...) {
	va_list args;
	va_start(args, format);
	schedfBufferPos += vsprintf(schedfBuffer + schedfBufferPos, format, args);
	va_end(args);
}

void flushschedf() {
	printf("%s", schedfBuffer);
	schedfBuffer[0] = '\0';
	schedfBufferPos = 0;
}

u32 g_testThreadDoneSema = 0;

extern int test_main(int argc, char *argv[]);

char testThreadStack[TEST_THREAD_STACK_SIZE] __attribute__ ((aligned(16)));
void testThreadProc() {
	//TODO: Print begin/end traces
	//TODO: Pass arguments from main thread to the test function
	test_main(0, NULL);
	SignalSema(g_testThreadDoneSema);
}

int main(int argc, char *argv[]) {
	//A thread is created to ensure the execution environment is the same
	//across all possible boot methods (direct hw boot, ps2link, emulator, etc.)
	
	ee_sema_t semaInfo;
	memset(&semaInfo, 0, sizeof(ee_sema_t));
	semaInfo.init_count = 0;
	semaInfo.max_count  = 1;
	semaInfo.option     = 0;
	g_testThreadDoneSema = CreateSema(&semaInfo);
	
	u32 testThreadId = 0;
	ee_thread_t threadParam;
	memset(&threadParam, 0, sizeof(ee_thread_t));
	threadParam.func               = (void*)&testThreadProc;
	threadParam.initial_priority   = TEST_THREAD_PRIORITY;
	threadParam.stack_size         = TEST_THREAD_STACK_SIZE;
	threadParam.stack              = testThreadStack;
	testThreadId = CreateThread(&threadParam);
	
	StartThread(testThreadId, NULL);
	WaitSema(g_testThreadDoneSema);
	
	return 0;
}
