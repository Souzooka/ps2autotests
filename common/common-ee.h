#ifndef PS2AUTOTESTS_COMMON_EE_H
#define PS2AUTOTESTS_COMMON_EE_H

// Defines for MSVC highlighting.  Not intended to actually compile with msc.
#ifdef _MSC_VER
#define __STDC__
#define _EE
#define __attribute__(x)
#endif

#include <stdio.h>
#include <tamtypes.h>

#undef main
#define main test_main

void schedf(const char *format, ...);
void flushschedf();

#endif
