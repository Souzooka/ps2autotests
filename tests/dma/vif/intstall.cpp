#include <common-ee.h>
#include <string.h>
#include <kernel.h>
#include "../dmaregs.h"
#include "../dmasend.h"
#include "../dmatags.h"
#include "emit_vifcode.h"
#include "vifunits.h"

static VIF::Unit* g_units[2] = { &VIF::Unit0, &VIF::Unit1 };

typedef void (*WriteIntCodeFunction)(VIF::Packet&);

void writeNopInt(VIF::Packet& vifcode) {
	vifcode.NOP(0, 0, VIF::CMD_ENABLE_INT);
}

void writeMarkInt(VIF::Packet& vifcode) {
	vifcode.MARK(0xBEEF, VIF::CMD_ENABLE_INT);
}

void writeStMaskInt(VIF::Packet& vifcode) {
	vifcode.STMASK(0xFFFFFFFF, VIF::CMD_ENABLE_INT);
}

void doTest(VIF::Unit* unit, const char* testName, WriteIntCodeFunction writeIntCode) {
	static const u32 vifPacketQwc = 3;
	static const u32 vifPacketSize = vifPacketQwc * 0x10;
	
	unit->regs->fbrst = VIF::FBRST_RST;
	
	VIF::Packet vifcode(vifPacketSize);

	vifcode.STMASK(0xAAAAAAAA);
	
	writeIntCode(vifcode);
	vifcode.STMASK(0xBBBBBBBB);
	
	vifcode.NOP();
	vifcode.NOP();
	vifcode.STMASK(0xCCCCCCCC);
	
	DMA::SendSimple(unit->dmaChannel, vifcode.Raw(), vifPacketSize);
	
	printf("%s:\n", testName);
	printf("  Status after transfer - MASK: %08x, STAT: %08x\n", unit->regs->mask, unit->regs->stat);
}

int main(int argc, char *argv[]) {
	printf("-- TEST BEGIN\n");
	
	for(u32 i = 0; i < 2; i++) {
		VIF::Unit* unit = g_units[i];
		printf("== VIF%d ==\n", i);
		doTest(unit, "NOP (I)", &writeNopInt);
		doTest(unit, "MARK (I)", &writeMarkInt);
		doTest(unit, "STMASK (I)", &writeStMaskInt);
		printf("\n");
	}
	
	printf("-- TEST END\n");
	return 0;
}
