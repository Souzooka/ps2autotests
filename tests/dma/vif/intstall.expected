-- TEST BEGIN
== VIF0 ==
NOP (I):
  Status after transfer - MASK: aaaaaaaa, STAT: 00000c02
MARK (I):
  Status after transfer - MASK: cccccccc, STAT: 00000840
STMASK (I):
  Status after transfer - MASK: ffffffff, STAT: 00000c02

== VIF1 ==
NOP (I):
  Status after transfer - MASK: aaaaaaaa, STAT: 00000c02
MARK (I):
  Status after transfer - MASK: cccccccc, STAT: 00000840
STMASK (I):
  Status after transfer - MASK: ffffffff, STAT: 00000c02

-- TEST END
